src/acf.rs

```diff
@@ -0,0 +1,213 @@
+use {ring_buffer, Frame, Sample};
+
+/// Iteratively applies the ACF (auto-correlation function) over a signal of sample `Frame`s.
+///
+/// The `current` method returns a slice of float `Frame`s. Each frame in the acf represents how
+/// strongly correlated the signal is with itself at each sample offset, over the given window size.
+///
+/// In order to properly calculate all of the auto-correlation values, the `next` method must be
+/// called with at least `window_size() + max_t` frames. Prior to that, the output acf values are
+/// undefined.
+pub trait Acf<F> where F: Frame {
+
+    /// Updates the calculated acf with `new_frame`. Returns the updated acf values. The `current`
+    /// method will return this same slice after the call to `next` completes.
+    fn next(&mut self, new_frame: F) -> &[F::Float];
+
+    /// The current values calculated using the ACF. This is a slice of float frames, with a length
+    /// equal to the value returned by the `max_t` method.
+    fn current(&self) -> &[F::Float];
+
+    /// The size of the window used to calculate the ACF.
+    fn window_size(&self) -> usize;
+
+    /// The number of frame offsets (where each offset is referred to as "t") for which to apply
+    /// the ACF. The acf slice returned by the `current` and `next` methods will be max_t() + 1 in
+    /// size.
+    fn max_t(&self) -> usize;
+}
+
+/// An implementation of the Acf trait that uses the "diff squared" to calculate correlation.
+///
+/// The typical auto-correlation algorithm is used, but instead of multiplying samples together
+/// and summing over the window at each T offset, the difference between the samples is squared
+/// over the window for each T offset, and those are summed together.
+#[derive(Clone)]
+pub struct DiffSquaredAcf<F>
+where
+    F: Frame,
+{
+    window_size: usize,
+    max_t: usize,
+    frame_buffer: ring_buffer::Fixed<Vec<F::Float>>,
```

Comment:
Here a `Vec` is used as the internal buffer for the fixed-size ring buffer, however it seems like we should be able to support `ring_buffer::Fixed<S>` over any type `S` where `S: ring_buffer::Slice`. I know I mentioned using `Vec` in the `DiffSquaredAcf` type as a workaround to #94, however if we are to add this to the `sample` crate itself I think we should remain generic over the inner data slice type (in case the user would like to re-use a buffer, a fixed-size array or some other custom buffer type). In this case, the type might look something like this:

```rust
pub struct DiffSquaredAcf<S>
where
    S: ring_buffer::Slice,
{
    window_size: usize,
    max_t: usize,
    frame_buffer: ring_buffer::Fixed<S>,
    acf: Vec<S::Element>,
}
```
I wonder if something similar could be done for the `acf` buffer itself too? It would be a pity to always require a `Vec` allocation when the user should probably be able to use a fixed-size array or re-use some other buffer if they like. Will have to keep thinking on this.

Reply:
No worries on the delay! I know how that happens.

I definitely considered that this is a dynamic allocation, and I know the rest of the library doesn't do that. I actually started out by having the user pass a ring_buffer:Fixed<S>. However, in this case, doing that makes the DiffSquaredAcf a bit awkward to use. That buffer needs to be a very specific size. Having the user create it is a bit awkward because it has to have a size of at least `max_t + window_size + 1`. Is there a clever way to hide that from the user?

Reply:
It may not be the end of the world to simply document clearly how big it needs to be, and also add a runtime assert to make sure it's large enough. What do you recommend?

Reply:
Panicking when the size is wrong is also what `interpolate::Sinc` does (it panics when the size is odd), so it'd be consistent.

Reply:
Sounds good. I'm using it as-is in my project for now. I'll come back to this PR when I have a chance to refactor my code.